stan/math/opencl/kernels/matrix_multiply.hpp

```diff
@@ -114,6 +114,94 @@ const local_range_kernel<cl::Buffer, cl::Buffer, cl::Buffer, int, int, int,
     matrix_multiply("matrix_multiply", matrix_multiply_kernel_code,
                     {{"THREAD_BLOCK_SIZE", 32}, {"WORK_PER_THREAD", 8}});
 
+// \cond
+static const char* matrix_vector_multiply_kernel_code = STRINGIFY(
+    // \endcond
+    /**
+     * Matrix-vector multiplication R=A*B on the OpenCL device
+     *
+     * @param[in] A matrix in matrix-vector multiplication
+     * @param[in] B vector in matrix-vector multiplication
+     * @param[out] R the output vector
+     * @param[in] M Number of rows for matrix A
+     * @param[in] N Number of cols for matrix A and number of rows for vector B
+     */
+    __kernel void matrix_vector_multiply(
+        const __global double* A, const __global double* B, __global double* R,
+        const int M, const int N) {
+      const int gid = get_global_id(0);
+
+      double acc = 0;
+      for (int i = 0, j = 0; i < N; i++, j += M) {
+        acc += A[j + gid] * B[i];
```

Comment:
You are leaning on GPUs own caching mechanism for memory reads right? Would we benefit with local memory here at all? This one is optional as this version is already faster, so we can leave that one for future optimizations.

Reply:
Right, but only for reading elements of B. I think we would likely only see benefit on older GPUs that do not cache reads from global memory.